dbms/src/Common/MPMCQueue.h

```diff
@@ -274,49 +283,47 @@ class MPMCQueue
     }
 
     template <bool need_wait>
-    bool popObj(T & res, [[maybe_unused]] const TimePoint * deadline = nullptr)
+    Status popObj(T & res, [[maybe_unused]] const TimePoint * deadline = nullptr)
     {
 #ifdef __APPLE__
         WaitingNode node;
 #else
         thread_local WaitingNode node;
 #endif
-        {
-            std::unique_lock lock(mu);
+        std::unique_lock lock(mu);
 
-            if constexpr (need_wait)
-            {
-                /// read_pos < write_pos means the queue isn't empty
-                auto pred = [&] {
-                    return read_pos < write_pos || !isNormal();
-                };
-                wait(lock, reader_head, node, pred, deadline);
-            }
+        if constexpr (need_wait)
+        {
+            /// read_pos < write_pos means the queue isn't empty
+            auto pred = [&] {
+                return read_pos < write_pos || !isNormal();
+            };
+            if (!wait(lock, reader_head, node, pred, deadline))
+                return Status::TIMEOUT;
+        }
+        if (!isCancelled() && read_pos < write_pos)
+        {
+            auto & obj = getObj(read_pos);
+            res = std::move(obj);
+            destruct(obj);
 
-            if (!isCancelled() && read_pos < write_pos)
-            {
-                auto & obj = getObj(read_pos);
-                res = std::move(obj);
-                destruct(obj);
-
-                /// update pos only after all operations that may throw an exception.
-                ++read_pos;
-
-                /// Notify next writer within the critical area because:
-                /// 1. If we remove the next writer node and notify it later,
-                ///    it may find itself can't obtain the lock while not being in the list.
-                ///    This need carefully procesing in `assignObj`.
-                /// 2. If we do not remove the next writer, only obtain its pointer and notify it later,
-                ///    deadlock can be possible because different readers may notify one writer.
-                notifyNext(writer_head);
-                return true;
-            }
+            /// update pos only after all operations that may throw an exception.
+            ++read_pos;
+
+            /// Notify next writer within the critical area because:
+            /// 1. If we remove the next writer node and notify it later,
+            ///    it may find itself can't obtain the lock while not being in the list.
+            ///    This need carefully procesing in `assignObj`.
+            /// 2. If we do not remove the next writer, only obtain its pointer and notify it later,
+            ///    deadlock can be possible because different readers may notify one writer.
+            notifyNext(writer_head);
+            return Status::NORMAL;
         }
-        return false;
+        return isNormal() ? Status::EMPTY : status;
```

Comment:
In which case will status be normal at here? Since before this pr, we always return false, which means the queue is cancel or finished.

Reply:
`tryPop` won't wait even if queue itself is normal, in such case we should return `EMPTY`.